byway/src/main/java/com/google/sps/servlets/DestinationsServlet.java

```diff
@@ -87,9 +89,42 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     }
     datastore.put(entity);
     ArrayList<String> destinations = (ArrayList<String>) entity.getProperty("destinations");
-    UserLocations userLocations = new UserLocations(start, destinations);    
+    Trip trip = new Trip(KeyFactory.keyToString(tripKey), start, destinations, new ArrayList<String>(), new ArrayList<String>());    
     response.setContentType("application/json;");
-    response.getWriter().println(gson.toJson(userLocations)); 
+    response.getWriter().println(gson.toJson(trip));
   }
-}
+  
 
+  /*
+  * Adds new user entity if user has not logged in before
+  * Adds tripKey to the user's array of trips
+  **/
+  public void addUserEntity(Key tripKey){
+    boolean userExists = false;
+    String userId = userService.getCurrentUser().getUserId();
+    String userEmail = userService.getCurrentUser().getEmail();
+    Query query = new Query("User");
+    PreparedQuery results = datastore.prepare(query);
+    for (Entity currentEntity : results.asIterable()) {
```

Comment:
Firstly, iterating through every user in the database like this is not practical in any real world application. You should sort through data via your query and setting filters to retrieve the data you want. However in this case, we don't need to do that as per my second point.

Second, as discussed in Leo's pull request, we can come up with a standardized way of creating keys for users. This way we can quickly look them up in the database. See this comment chain:
https://github.com/googleinterns/step200-2020/pull/19#discussion_r466764313

Third, this method is doing too much and can be broken up. It makes more sense to have two separate methods, one to get/add the user and another to add the trip.

So assuming we switch to the approach Zach outlines, you can update your logic like this:
```java
public UserInfo addUserEntity(User user) {
  // Create a key based on the user ID
  Key userKey = KeyFactory.createKey(UserInfo.KIND, user.getUserId());
  UserInfo userInfo;
  try {
    // try to retrieve the entity with the key 
    Entity userEntity = datastore.get(userKey);
    userInfo = UserInfo.FromEntity(userEntity);
  catch (EntityNotFoundException exception) {
    // If the user doesn't exist yet or is new, create a new user
    Entity newUserEntity = new Entity(userKey)
    datastore.put(newUserEntity);
    userInfo = UserInfo.FromEntity(newUserEntity)
  }
  return userInfo;
}

public void addTripForUser(UserInfo user) {
  // left as an exercise for the reader ;)
  ...
}
```

It also may be better to create new UserInfo on the login section instead, but either will work.

Reply:
fixed it